### PR TITLE
Salt-ssh can deploy a relenv environment

### LINKED ISSUE #66877: [FEATURE REQUEST] Add `--relenv` Option to `salt-ssh` for Using a Onedir Bundled Version of Salt+Python
**Summary:**

This feature request proposes adding a `--relenv` option to `salt-ssh` that allows downloading and using a bundled version of Salt and Python (relenv) on the SSH target, similar to how the `--thin` option works. This will serve as a transitional feature towards making `relenv` the default method for Salt-SSH deployments in the future.

**Description:**

Currently, `salt-ssh` uses the `--thin` option to copy a minimal version of Salt to the SSH target, along with Python dependencies, and then run Salt remotely. However, with the increasing preference for using `relenv` bundled environments (Salt and Python packaged together), it would be ideal to introduce a `--relenv` option that downloads and uses a one-directory (onedir) version of Salt+Python on the target machine.

The `--relenv` option would follow a similar flow as `--thin`, but instead of copying a minimal installation, it would download and unpack a `relenv` bundle (which includes both Salt and Python). This could be useful for environments where managing Python versions and dependencies on the target machine is difficult or where a fully isolated environment is desired.

### Expected Behavior:

When a user runs `salt-ssh` with the `--relenv` option, Salt should:

1. **Check for `relenv` support on the target**: Ensure that the target system can support the `relenv` bundle.
2. **Download the `relenv` bundle**: Retrieve the pre-packaged Salt+Python bundle (possibly from a pre-configured URL or file server).
3. **Extract and use**: Extract the bundle on the target and use it to run Salt commands.
4. **Handle clean-up**: Optionally clean up the environment after execution (if the user specifies this).
  
This option will allow users to run Salt on machines that may not have Salt or Python pre-installed or where version conflicts exist.

**Proposed CLI Syntax:**

```bash
salt-ssh --relenv <target> test.ping
```

Additional options for managing the `relenv` bundle can include paths to the local or remote onedir package, force re-download, etc.

**Justification:**

The `relenv` bundle represents the future direction for Salt deployments, and adding this as an optional feature in `salt-ssh` will enable users to transition smoothly to the new system. It also simplifies Salt-SSH deployments on targets that have conflicting or missing Python versions and dependencies.

**Current Behavior:**

The current `salt-ssh` approach uses `--thin` to transfer a lightweight Salt environment. This requires Python to be present on the target and can run into compatibility issues depending on the target’s system configuration.

**Additional Context:**

- Once stabilized, `relenv` can eventually become the default for `salt-ssh`.
- This facilitates using salt-ssh to bootstrap `salt` on the targets
- This will facilitate the deployment of ephemeral minions with `salt-ssh`

### PR DESCRIPTION
**Pull request by Akm0d**
### What does this PR do?
This PR adds a new `--relenv` option to `salt-ssh` that deploys a relenv (onedir) bundled version of Salt and Python on the SSH target. The `--relenv` option works similarly to `--thin` but deploys a bundled Salt environment for cases where it’s preferable to use relenv.

### What issues does this PR fix or reference?
Fixes [#66877](https://github.com/saltstack/salt/issues/66877)

### Previous Behavior
`Salt-ssh` only supported the `--thin` option, which deploys a minimal Salt environment. 

### New Behavior
The new `--relenv` option allows `salt-ssh` to deploy a full Salt environment bundled with Python, offering more flexibility in environments where relenv is preferable. This will eventually become the new default, but for now, it remains an option.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes

### CODE REVIEW

File: salt/client/ssh/__init__.py [hunk 1]
```diff
@@ -1141,14 +1142,140 @@ def check_thin_dir(self):
...
+        # Parse the output
+        try:
+            kernel, arch, winarch = stdout.lower().strip().split("|", maxsplit=2)
+        except ValueError as e:
+            log.error(f"Error parsing OS/arch detection result: {e}")
+            raise ValueError("Failed to parse OS and architecture data")
+
+        # Set architecture
+        if "arm" in arch:
+            os_arch = "arm64"
```
Comment (dmurphy18) [inline salt/client/ssh/__init__.py:1168]: Wondering about MacOS  arm64, and Debian/Ubuntu aarch64 ??

```
ubuntu@ip-10-2-5-97:~$ uname -p
aarch64
ubuntu@ip-10-2-5-97:~$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 24.04 LTS
Release:	24.04
Codename:	noble
ubuntu@ip-10-2-5-97:~$ 
```

**Review by dmurphy18 — COMMENTED** [review]
Don't forget the changelog entry, and tests.

File: salt/client/ssh/__init__.py (hunk 1 above)
Comment (Akm0d) [reply salt/client/ssh/__init__.py:1168]: I don't care about matching *every* architecture perfectly, only the ones that have tarballs in our latest onedir repo:
https://repo.saltproject.io/salt/py3/onedir/latest/

MacOS arm64 is covered by this code.
We only have a linux tarball for `x86_64` and `arm64`, so this code would match aarch64 to x86_64

Comment (Akm0d) [reply salt/client/ssh/__init__.py:1168]: oh i see, `aarch64` needs to match `amd64`

Comment (dmurphy18) [reply salt/client/ssh/__init__.py:1168]: Nope, Linux uses ```aarch64``` for Arm architectures, Apple uses ```arm64``` for Arm Architectures.
```amd64``` is equivalent to Intel x86 64 bit

**Comment by twangboy** [comment]
Failing lint and pre-commit

**Review by dmurphy18 — DISMISSED** [review]

File: tests/pytests/integration/ssh/test_deploy.py [hunk 2]
```diff
@@ -163,6 +163,28 @@ def test_thin_dir(salt_ssh_cli):
...
+    assert thin_dir.joinpath("salt-call").exists()
+
+
+def test_relenv_ping(salt_ssh_cli):
+    """
+    Test a simple ping
+    """
+    ret = salt_ssh_cli.run("--relenv", "test.ping")
+    assert ret.returncode == 0
+    assert ret.data is True.joinpath("salt-call").exists()
```
Comment (twangboy) [inline tests/pytests/integration/ssh/test_deploy.py:185]: Something's not right here... copy/paste error?

**Review by twangboy — CHANGES_REQUESTED** [review]

**Review by twangboy — APPROVED** [review]

**Review by dmurphy18 — APPROVED** [review]

**Pull request merged**